infrastructure/lib/constructs/lambda-fn.ts

```diff
@@ -0,0 +1,219 @@
+      // editing name from API spec
+      (this.fn.node.defaultChild as lambda.CfnFunction).overrideLogicalId(id + "Function");
+    }
+
+    // Optional - create DynamoDB connection and grant permissions
+    if (props.table && props.tablePermissions) {
+      this.table = props.table;
+      this.fn.addEnvironment("ATAT_TABLE_NAME", props.table.tableName);
+      this.grantTablePermissions(props.tablePermissions);
+    }
+
+    // Optional - grant permissions for secrets manager
+    if (props.smtpSecrets) {
+      props.smtpSecrets.grantRead(this.fn);
+    }
+
+    // Optional - create S3 bucket connection and grant permissions
+    if (props.bucket && props.bucketPermissions) {
+      this.bucket = props.bucket.bucket;
+      this.fn.addEnvironment("DATA_BUCKET", this.bucket.bucketName);
+      this.grantBucketPermissions(props.bucketPermissions);
+    }
+
+    // Optional - create SQS connection and grant permissions
+    if (props.queue && props.queuePermissions) {
+      this.queue = props.queue;
+      this.fn.addEnvironment("ATAT_QUEUE_URL", props.queue.queueUrl);
+      this.grantSQSPermissions(props.queuePermissions);
+      // allows the fn to subscribe to the queue using an EventSource
+      if (props.createEventSource) {
+        this.fn.addEventSource(new SqsEventSource(this.queue, props.batchSize ? { batchSize: props.batchSize } : {}));
+      }
+    }
+
+    // Optional - create State Machine and grant permissions
+    if (props.stateMachine) {
+      this.stateMachine = props.stateMachine;
+      this.fn.addEnvironment("SFN_ARN", props.stateMachine.stateMachineArn);
+      this.stateMachine.grantStartExecution(this.fn);
+    }
+  }
+
+  private grantTablePermissions(tablePermissions: string | undefined): iam.Grant | undefined {
+    switch (tablePermissions) {
+      case "ALL":
+        return this.table.grantFullAccess(this.fn);
+      case "READ":
+        return this.table.grantReadData(this.fn);
+      case "READ_WRITE":
+        return this.table.grantReadWriteData(this.fn);
+      case "WRITE":
+        return this.table.grantWriteData(this.fn);
+      default:
+        return this.table.grantReadWriteData(this.fn);
+    }
+  }
+
+  private grantBucketPermissions(bucketPermissions: string | undefined): iam.Grant | undefined {
+    switch (bucketPermissions) {
+      case "READ":
+        return this.bucket.grantRead(this.fn);
+      case "PUT":
+        return this.bucket.grantPut(this.fn);
+      case "DELETE":
+        return this.bucket.grantDelete(this.fn);
+      default:
+        return this.bucket.grantRead(this.fn);
+    }
+  }
+
+  private grantSQSPermissions(queuePermissions: string | undefined): iam.Grant | undefined {
+    switch (queuePermissions) {
+      case "SEND":
+        return this.queue.grantSendMessages(this.fn);
+      case "CONSUME":
+        return this.queue.grantConsumeMessages(this.fn);
+      default:
+        return this.queue.grantSendMessages(this.fn);
```

Comment:
Is it really appropriate to just grant `send` permissions on any invalid value? Why is that not an error?

Reply:
Yeah, agree not a great default behavior upon viewing this with fresh eyes.